packages/interaction/README.md

```diff
@@ -0,0 +1,302 @@
+# @bigtest/interaction
+
+Synchronously interact with DOM elements using a
+[convergence](https://github.com/thefrontside/bigtest/tree/master/packages/convergence)
+interface.
+
+## Convergent Interactions
+
+This libarary uses convergences from
+[`@bigtest/convergence`](https://github.com/thefrontside/bigtest/tree/master/packages/convergence)
+to ensure that elements exist in the DOM before being interacted
+with. As such, the `Interaction` class provided by this library
+supports the `Convergence` methods  as well as interaction specific
+ones.
+
+``` javascript
+import { Interaction } from '@bigtest/interaction';
+
+let logIn = new Interaction('#login-form')  // optional scope selector
+  .fill('.email-input', 'email@domain.tld') // fills in an email
+  .blur('.email-input')                     // triggers blur validation
+  .fill('.pass-input', '5up3rS3cr37')       // fills in a password
+  .click('.submit-btn')                     // submits the form
+  .once(() => !!user.loggedIn)              // ensures we've logged in
+
+// runs the interaction and bails if it cannot complete in 1 second
+logIn.timeout(1000).run()
+  // .run() returns a promise
+  .then(() => {...})  // user has been logged in
+  .catch(() => {...}) // something went wrong along the way
+
+// interactions are immutable and can be run multiple times
+logIn.do(() => console.log('logged in')).timeout(500).run()
+```
+
+### Available Interaction Methods
+
+- `.click(selector)`
+- `.fill(selector, value)`
+- `.focus(selector)`
+- `.blur(selector)`
+- `.trigger(selector, event, eventOptions)`
+- `.scroll(selector, { left, top })`
+
+These interaction methods can omit the `selector` and interact
+directly with the scoped element of the interaction.
+
+In addition to the above, you may use the `.find(selector)` method
+to converge on an element existing. That element is then passed on to
+the next convergence method in the stack.
+
+``` javascript
+new Interaction('#login-page')
+  .find('h1.heading')
+  .once((h1) => h1.innerText === 'Log In')
+  .do(() => console.log('looks good!'))
+  .timeout(100)
+  .run()
+```
+
+Similarly, there is `.findAll(selector)`, but this method does not
+converge on elements existing. Instead it will return an empty array
+if it cannot find any elements matching `selector`.
+
+### Custom Interaction Methods
+
+You can register custom interaction methods and use them in new
+`Interaction` instances.
+
+``` javascript
+Interaction.register('logIn', function(email, password) {
```

Comment:
If we want to ditch `Interaction.register`, we could actually just extend the Interaction for custom ones.

This works currently:

```javascript
class CustomInteraction extends Interaction {
  login(email, password) {
    return this
      .fill(...)
      .fill(...)
  }
}
```

One of the things that this doesn't quite solve is the future `visitable` helper. My original idea is that in your app setup, you could just do `Interaction.register('visit, () => {...})` and all `visitable` helpers would use that registered `Interaction.visit` method.

But there are probably some other concerns there as well; such as creating a `new Interaction` _before_ the `register('visit', ...)` happens. In that case, it will not have access to the visit helper.

Could there be some other easy-entry way to attach a history interface to the visitable helper? Especially in React apps, we can't access the history API unless the app is mounted first (in a hook).

Reply:
If we make it extensible, then we can just extend multiple times. Once when we're defining the "global" application interactions, and then again once the router is available:

```js
// tests/helpers/interactions.js

class ApplicationInteraction extends Interaction {
  login(email, password) {
    return this.
      fill(...)
      .fill(...)
  }
}
```

Then later on in the test harness....

```js
// tests/helpers/harness.js

//....
 componentWillMount() {
    let history = createMemoryHistory();
    this.Interaction = class NavigableInteraction extends ApplicationInteraction {
      visit(path) {
        history.push(path);
    }
 }
//...
```

The testcases would use the `TestHarness`'s interaction class to perform the actual interactions.

The main thing that strikes me as positive about this approach is that we're basically using the prototype chain of the `Interaction` hierarchy to aggregate the interactions available to the application. This is immutable, and composable, and so if we want to layer on a more function-oriented api with mixins and factories and what not that hide the fact that it's all based on prototypes, that can be done easily.